### ISSUE TITLE
On supporting goroutines

### ISSUE THREAD

**Issue by PXshadow** [go2hx/go2hx#190]
Goroutines is a mixture of threading (if available) and coroutines.

In the case of Haxe, threading is supported on all targets except js, but there is no native coroutine support.
The most straight forward approach is to treat every goroutine as one thread for all system targets.

As a next step, coroutines could be included by ast rewriting/analysis on the [go side](https://github.com/go2hx/go2hx/tree/master/analysis) by writing in places for yeilds along with other markers to support a coroutine controlflow.

The other piece is having per goroutine panic and recover logic as covered in #187 
It would be wise to get double panicing working outside of a goroutine context, and then once working, support it with goroutines.

The final step would be a combination of both approaches, similar to how go handles goroutines internally, with an algorithm to group multiple goroutines on a single thread and use coroutines, to be able to support thousands of goroutines, with a limited amount of threading being shared acrossed. Coroutine handling controlling the jumping between all of the goroutines execution across parallel execution environments seamlessly. 

This level of support should only be considered, once the compiler is fully spec compliant, and most problems are now related to performance or wanting to support things out of spec.

Feel free to express your thoughts on this, and let's turn this into an open discussion. A discussion on approaches for implementations, scoping, and timeline.

**Comment by elliott5** [comment]
I'll come back with a more considered response to the overall topic in a day-or-so.

Meantime, last week I went to see how GopherJS does it and put the link below aside for when the topic came up. Unfortunately my JS is not good enough to understand how it works, but I leave it here to fuel the debate for those more competent than I.

[GopherJS goroutine implementation](https://github.com/gopherjs/gopherjs/blob/master/compiler/prelude/goroutines.js)

**Comment by elliott5** [comment]
Apologies for my delayed response, I agree with your analysis @PXshadow. 

There are two approaches to implementing goroutines:
- Either use a target specific approach (threads for Haxe Sys targets, GopherJS approach or similar for JS); or
- Use a generic approach.

The problem I see with multiple threads, is that users of our generated Haxe library my be surprised to find it creating additional threads that they were not expecting. Especially if some of those threads persist after the call to the library.


### The generic approach

I can only recount what I did with TARDISgo... other approaches my be equally valid.

This requires each goroutine to have its own goroutine-stack. 

A goroutine-supervisor loop runs each goroutine as required. TARDISgo just ran each goroutine in-turn, but more complex strategies based on which goroutines are waiting on which channels would be more efficient.


### Goroutine functions - a function called in the context of a goroutine

Goroutine functions can be created by extending the existing goto go-rewriting code.

Goroutine functions have a calling signature that has the goroutine-stack as the first parameter.

Goroutine functions keep all of their local variables in a stack-frame on the goroutine-stack, along with their next goto-location (as in the current goto implementation). There may also be some defer/panic/recover data in the stack-frame.

They execute in a loop, in the same way as goto re-written functions do now.

When context-switching occurs (usually with a channel request) the goroutine saves its next goto-location alongside the other local variables on the goroutine-stack and returns to the goroutine-supervisor, which then runs the next goroutine.

When a goroutine restarts, it simply executes from the stored next goto-location.


### Non-goroutine functions (a possible future optimisation)

TARDISgo was very simple, and only had the goroutine functions described above.

A simple optimisation would be to observe that not all functions need to run in the context of a goroutine.

Functions that do not do any context-switching, and do not call any other functions that do any context-switching can be implemented as they are now.

Static analysis across the whole library would be required to maximise the number of non-goroutine functions identified.

However, panic/defer/recover issues would need to be resolved.

**Comment by elliott5** [comment]
I should say, when a pure Haxe program calls into a go2hx library, a new goroutine would need to be created to run it, then destroyed after the call is over. This could be done by the go2hx API functions.

**Comment by PXshadow** [comment]
How would a goroutine be created from the haxe side? Considering a goroutine for go needs to have the AST rewritten (in the same manner as goto)

I appreciate you laying out in detail how this can be achieved, the context switching transformations make sense, and given how the generic approach allows support of all Haxe targets, it seems quite advantageous, including as you pointed out the issue of potential unexpected threads being created, and the potential for them not to be closed.

So the timeline in my head looks to be the following:

* Fix the panic/defer/recover issues
* Look over the internals of channels and fix any current short comings
* Get the thread only implementation working with no known issues
* Re-include all Chan/goroutines tests for the Go and Yaegi tests (easy,med) and pass them all.
* Create the generic approach using the same technique as goto transformations, with some type of basic marking system to know when to create the context switching transformations + stack variables etc.
* Get the generic only implementation working with no known issues.
* Pass the chan/goroutines tests using the generic approach.

**Comment by elliott5** [comment]
# Goroutine-enabled function calls

First we need to discuss how one goroutine-enabled function calls another.

The whole call/return process has to be simulated on the goroutine-stack.

Note that one goroutine-enabled function never calls another goroutine-enabled function directly using the Haxe stack, the goroutine-scheduler (described briefly below) performs this function.

Each goroutine-enabled function has a stack-frame that contains all its data. 

The Haxe call signature of all goroutine-enabled functions is the same, a single parameter referencing the stack-frame. (In future, there could also be return values that relate to optimising the goroutine scheduling process and handling panic/defer/recover processing.)

In TARDISgo this was a Haxe type that was a struct containing:
- the call parameters
- the return parameters
- the local variables
- reference to the Haxe function containing the code (so that the goroutine-scheduler knows what to run)
- the goto-location to start when the function is next executed
- probably some panic/defer/recover data
- possibly some debug info, for creating a stack-trace

Consider when goroutine-enabled function A "calls" goroutine-enabled function B, which then "returns".

Goroutine-enabled function A "call":
- creates a stack-frame for function B 
- fills in function B's call parameters
- sets function B's goto-location to 0 (meaning run the function B initialisation code)
- saves the function B stack-frame in local variable R
- pushes the function B stack-frame onto the goroutine-stack (still also referenced from R)
- sets function A's own goto-location as the code to execute next after the call to function B ends
- return to the goroutine-scheduler, so that it can run function B with the created stack-frame (more on this anon)

Goroutine-enabled function B (run by the goroutine-scheduler) "execution/return":
- sees that the goto-location is 0, so run the function initialisation code
- runs the logic of the program as normal (current goto implementation)
- fills the return values into its own stack-frame.
- "returns" by popping its own stack-frame off the goroutine-stack, then returning control to the goroutine-scheduler.

Goroutine-enabled function A (run by the goroutine-scheduler) "return admin":
- restarts at the goto-location to handle the post-call admin
- uses local variable R to extract the return values from the function B "call" into local variables
- sets variable R to be null, so that the function B stack-frame is garbage collected
- carries on program execution as now...

# Simple goroutine-scheduler

All the goroutines are represented by an array of goroutine-stacks. Array element 0, contains the main goroutine.

Every time the goroutine-scheduler is run from Haxe:
- It loops over the array to examine each goroutine-stack in turn.
- If the examined goroutine-stack is not empty, execute the goroutine-enabled function using the top stack-frame. 
- When the main goroutine stack is empty (element 0), return from the goroutine-scheduler to the Haxe caller.

# How to execute a "go" from a goroutine-enabled function

Create a stack-frame for the goroutine-enabled function to be "go"ed (as described above).

Add another element to the goroutine-stack array.

Push the stack-frame onto the new goroutine-stack in the array.

Maybe give up control to the goroutine-scheduler, to allow the new goroutine to be executed.

# How to run a goroutine-enabled go2hx function from pure Haxe

Create a stack-frame for the goroutine-enabled go2hx function (as described above) and put it in local variable R.

Push the stack-frame onto the goroutine-stack at element 0 of the goroutine array (the main goroutine).

Call the goroutine-scheduler.

Handle the goroutine-enabled function "return admin" (as described above) using local variable R.

**Comment by PXshadow** [comment]
Great video on how to go from a simple Goroutine scheduler to the one Go uses: https://youtu.be/-K11rY57K7k?si=Zbr7tJFveIKPYnlZ

**Comment by elliott5** [comment]
If the go2hx goroutine scheduler has 10% of the functionality described in this video I will be delighted!

**Comment by PXshadow** [comment]
Nice document describing the runtime with mention of the Go scheduler: https://github.com/golang/go/blob/c386ed107e65f495acb6aacb0efcaade27da6a15/src/runtime/HACKING.md?plain=1#L6

**Comment by elliott5** [comment]
The good news is, with your excellent internal documentation spot @PXshadow, we have a pretty comprehensive cook-book from the Go team to take a set of goroutines implemented as co-routines and build a scheduler for them that works across multiple threads...

...all we need now are goroutines implemented as co-routines inside the go2hx implementation - as discussed earlier above.

**Comment by TheTechsTech** [comment]
I haven't got into the details of https://github.com/RealyUniqueName/Coro, https://github.com/nadako/haxe-coroutines, https://lib.haxe.org/p/yield/, https://lib.haxe.org/p/tink_await/, have they been tested/considered for general construction? 

It seems they have a better handle on how to process.

**Comment by elliott5** [comment]
I think we're hoping the main Haxe compiler will ride to our rescue and choose a coroutine implementation that we can just use off the shelf @TheTechsTech.

**Comment by PXshadow** [comment]
Nadako's implementation is what the current PR for the Haxe compiler one is based off of.  The others have been looked out but didn't fit for various reasons. The coroutine that we are hoping to come from the main Haxe compiler is planned, but is still not fully fleshed out. The original writer of it, is busy with other things, and Simon has been left with the ball to push to the finish line, but he has run into various technical problems, if you read the PR comments. If there is one thing that Simon is, he's persistent, but I fear it may be delayed from lack of a clear vision and implementation.

**Comment by elliott5** [comment]
The compromise approach, if not all Haxe or all Go, would be for this project to use and maintain a public Haxe co-routine library, then extend that with a threaded scheduler. With starting options as outlined by @TheTechsTech [above](https://github.com/go2hx/go2hx/issues/190#issuecomment-2776235239).

As my programming experience is primarily in Go, I'm not the best person to choose which existing Haxe co-routine library provides the best starting point. 

Perhaps an approach would be to speak to the authors of these existing Haxe co-routine libraries to see if any of them are interested in collaborating to create a Haxe multi-threaded co-routine library that would both implement goroutines for go2hx and be of service to the wider Haxe ecosystem.

**Comment by TheTechsTech** [comment]
>  this project already simulates goroutines by using a thread per goroutine.

Not close, really glossing over the subject, the [goto.go](https://github.com/go2hx/go2hx/blob/3b321146cd1f02fc925dbbde1cc1684063ce8434/analysis/goto.go#L18) does not capture any state, does not do any **context** switch, and can't, seems to be regular function expression handling. 

Maybe point to something **specific**, that has **`go`** prefixed in **front** of function handling that.

To really understand the **Go scheduler** video you posted at the top, you need some historical perspective, what is gliding the whole talk, and how others should proceed.

[CppCon 2015: Pablo Halpern "Work Stealing"](https://www.youtube.com/watch?v=iLHNF7SgVN4)

[MIT 6.172 - Performance Engineering of Software Systems](https://www.youtube.com/playlist?list=PLUl4u3cNGP63VIBQVWguXxZZi0566y7Wf) at [#13](https://youtu.be/Z7r4aAZ9Vqo?si=rOIInTQAF2xfUMEE)

[Stanford CS149 - Parallel Computing](https://www.youtube.com/playlist?list=PLoROMvodv4rMp7MTFr4hQsDEcX7Bx6Odp) at [#5](https://youtu.be/mmO2Ri_dJkk?si=IGVhFvlQ8YEO3Vv3)

**Comment by elliott5** [comment]
You are right @TheTechsTech, I have glossed over the subject, the state that [goto.go](https://github.com/go2hx/go2hx/blob/3b321146cd1f02fc925dbbde1cc1684063ce8434/analysis/goto.go#L18) does currently capture are the local variables (which need to go inside the call-frame on a simulated goroutine stack, along with the call and return parameters) and the `LoopLabelMap` index of `token.Pos` which provides a program-counter with which to resume computation when returning from yielding control. My only defence is that over a decade ago I used this same method as a basis to simulate goroutines in Haxe in my [TARDISgo project](https://github.com/tardisgo/tardisgo). However the go/co-routine scheduler I created for that project was extremely basic, and single threaded. 

The overall objective of this project is to help to extend the Haxe ecosystem. So using this project's goroutine requirement to act as a catalyst for the creation of a Haxe library to provide coroutines, with a scheduler that can use multiple threads (see [gomaxprocs](https://pkg.go.dev/runtime#GOMAXPROCS)), would be a better outcome for the Haxe ecosystem than following what I did 10 years ago.

You clearly have more insight into how to create the Go scheduler than I do. So I hope you would be interested in contributing to a  Haxe threaded coroutine library. How do you think we should start such a library project?

**Comment by elliott5** [comment]
I notice that both [nadako/haxe-coroutines](https://github.com/nadako/haxe-coroutines?tab=readme-ov-file#other-computation-abstractions) (that forms the basis of the Haxe compiler coroutine development branch), and [RealyUniqueName/Coro](https://github.com/RealyUniqueName/Coro) generate code that uses the same underlying "State Machine per coroutine function" architecture as I did all those years ago in TARDISgo. 

Whereas [haxetink/tink_await](https://github.com/haxetink/tink_await) uses the promise/call-back paradigm.

**Comment by TheTechsTech** [comment]
It seems the problem is not the **code**, the person who has the vision, is the one capable, using the language they capable in. That can not be assigned. 

I think you should look back into **TARDISgo** maybe rewrite based on reviewing [proc.go](https://github.com/golang/go/blob/master/src/runtime/proc.go) it's doing as **MIT** and **Stanford** teaches. My [c-raii](https://github.com/zelang-dev/c-raii) does similar, **work stealing** disabled, buggy.

[RealyUniqueName/Coro](https://github.com/RealyUniqueName/Coro) seems more suitable, **the state machine**. You might need to listen to those lectures I listed, there are concepts that needs revisiting. There is a big difference being a **developer** to **programmer**, not interchangeable, one is a small aspect.  

I follow quite many projects, lately https://github.com/ziglang/zig/issues/6025 https://github.com/ziglang/zig/issues/23446 discussions they have similar goals, most my ideas i explore are coming directly from there, but they too fixed on terminology, putting a halt on progress.

**Comment by elliott5** [comment]
These are wise words @TheTechsTech, thank you. 

If the project cannot find a suitable pure-Haxe approach to goroutines, I will certainly asist the project to simulate goroutines in the TARDISgo manner.  Your pointers to Go's [proc.go](https://github.com/golang/go/blob/master/src/runtime/proc.go) and your [c-raii](https://github.com/zelang-dev/c-raii) are helpful in terms of the scheduling logic. I also note that your [c-coroutine](https://github.com/zelang-dev/c-coroutine) project is the most sophisticated management wrapper around [libuv](http://docs.libuv.org/) imaginable. 

Also helpful is your highlighting of [RealyUniqueName/Coro](https://github.com/RealyUniqueName/Coro) as the most suitable Haxe starting point. I will do more research, as you suggest. 

As you are monitoring a number of projects in this space, across different languages, I hope you will continue to monitor go2hx and offer further wise-council as the project matures.

**Comment by neimanpinchas** [comment]
While the project is talking more about existing projects, and not from unborn theories, I think it might be appropiate to share here what my toughts were when reviewing the various haxe couroutines attempts.

I felt that all of them (except coro wich I wasn't able to test and eximine, because its not pure haxe) are combining the scheduling code with the problem logic code, this makes the whole thing more complicated.

My approach would be to replace the sunspendable call flow with an array of OPS, then have a general purpose FORTH like vm, that would process that code, the VM should maintain its own multi task amangement, for each task an execution stack and a local varaible frame stack (sorry I told FORTH but I meant the PAUSE keyword).

From within the regular haxe side, calling a async fn would mean, adding a task to the VM, blocking on it would mean blocking on a sys.thread.Lock().

The VM should support these OPS
EXEC_NATIVE to run the next block
EXEC_CORO to push a pointer to another coroutine on top of execution stack
RETURN pop off from execution stack/callie should leave the return value(s) on value stack. optionally process panics by checking type
SET/GET to set locals
SETG/GETG to set globals
PAUSE to cooperatively pass to next task

anytime no task can be resumed, the VM should use a timer to rescedule  check.

I think the machine is simple enough and eaisy to visualize and could run on any of the haxe targets wihtout issue.

**Comment by TheTechsTech** [comment]
> I felt that all of them (except coro wich I wasn't able to test and eximine, because its not pure haxe) are combining the scheduling code with the problem logic code, this makes the whole thing more complicated.

@neimanpinchas You have PR on those projects fixing them? You have your pure Haxe implementation ready? Where is the link to the scheduling logic and problem logic mixup in the project? That should be a starter, to get things rolling here. Things need to start somewhere.

**Comment by neimanpinchas** [comment]
I made some attempts, while I learned from them, I have'nt yet found the reliable rocket to reach the moon.

I've attempted to open an issue with hscript, which has been quickly turned down due to being out of scope of project.

I have some local attempts, but I think they are too far from what I wrote above, personally I would start again from scratch, the above is actually a result from rethinking my previous implementations.

> Where is the link to the scheduling logic and problem logic mixup in the project?

I mean in general the fact of trying to make a state machine that must match the functionality of the async function, including all kinds of loops, branching suspending calls in arguments, in binops etc.

Currently I am using pecan, debugging is not easy, I hope to get something better ASAP.

**Comment by elliott5** [comment]
Like you, @neimanpinchas, I have been researching which Haxe "coroutine" libraries might be viable for use within go2hx. As you say, unfortunately [RealyUniqueName/Coro](https://github.com/RealyUniqueName/Coro) contains an Ocaml plugin, which is beyond my level of technical competence. And the github version of [nadako/haxe-coroutines](https://github.com/nadako/haxe-coroutines?tab=readme-ov-file#other-computation-abstractions) does not work with the latest version of Haxe. However, in good news, as you have spotted, [Aurel300/pecan](https://github.com/Aurel300/pecan/) seems to work and is available on TryHaxe, so it must be quite stable.  [My first  try](https://try.haxe.org/#245848e5) is [the pecan "getting started" example](https://aurel300.github.io/pecan/intro-start.html), but with 2 "instance" state-machines running side-by-side.

As has been discussed on this thread, while it is no small feat to convert an individual function into a state machine, those state machines need to orchestrated together in order to provide thread-like coroutines, running many nested state machine functions on a simulated stack, and then further integrated with OS functionality in order to work efficiently. 

I will do more research on pecan in this direction, and learn for myself the debugging difficulties. If you do find something better @neimanpinchas please pass it on.

**Comment by elliott5** [comment]
In my ideal world @neimanpinchas , this project would create a coroutine library which allowed your [reflaxe_go](https://github.com/neimanpinchas/reflaxe_go) project to generate "go my_haxe_translated_to_go_fn()" from existing Haxe code.

**Comment by elliott5** [comment]
For information, a go2hx branch has been created of [ohmrun/stx_coroutine](https://github.com/ohmrun/stx_coroutine/tree/go2hx), which certainly looks very interesting.

**Comment by TheTechsTech** [comment]
Maybe look into **Python** introduction of **yield**, [Demystifying Yield in Python](https://aaronnotes.com/2023/04/demystifying-yield-in-python/), how it evolved. 

From what I'm not seeing here, the feature of using closures. But that closure is just put into a queue, not actually executed at that moment. A switch is made to scheduler, the other coroutine, to take off queue. It seems nobody realize that is actually happening. 

What is the process in **Haxe** to have a variable hold a closure? Then pass that to pecan to execute, until you can do that, you will get nowhere near Goroutine.

**Comment by elliott5** [comment]
@neimanpinchas wrote above:

> My approach would be to replace the sunspendable call flow with an array of OPS, then have a general purpose FORTH like vm, that would process that code, the VM should maintain its own multi task amangement, for each task an execution stack and a local varaible frame stack (sorry I told FORTH but I meant the PAUSE keyword).

How would you generate these OPS? Using [reflaxe](https://github.com/SomeRanDev/reflaxe) or macros?

**Comment by neimanpinchas** [comment]
> How would you generate these OPS? Using [reflaxe](https://github.com/SomeRanDev/reflaxe) or macros?

I was thinking from build macros, that way I could cover the most targets, pecan claims that he had issues when he was working on the untyped AST, but i do not like the API.


Just recalled that https://github.com/profound7/moon-core?tab=readme-ov-file#generators is also clean, its just a huge dependicy, I wish I could split off the coroutine part.

**Comment by neimanpinchas** [comment]
> How would you generate these OPS? Using [reflaxe](https://github.com/SomeRanDev/reflaxe) or macros?

Now thinking of your question, can reflaxe be used as an alternative to build macros, for haxe preprocssing?

**Comment by PXshadow** [comment]
## My thoughts

> I think that the best outcome for the Haxe ecosystem as a whole would be for this project to act as a catalyst for https://github.com/HaxeFoundation/haxe/pull/11554 into the compiler, and then work with the compiler team to write a Go-style scheduler to allow multiple threads to execute those coroutines, as described https://github.com/go2hx/go2hx/issues/190#issuecomment-2735646603.
@elliott5 

Why don't we go down this route first and see where it goes?
It seems like we are getting distracted by theory and so many prior implementations.


We should work on an implementation of both coroutines and the scheduler in the compiler, leading to:
1. Best debugging support.
2. Best interoperability with other Haxe code. (go2hx can simply invoke the Haxe stdlib for coroutines+threads)
3. Win-Win scenario, Haxe coroutines supported, go2hx is able to support goroutines on a sincere fashion, go2hx provides a large test suite and benchmarks for goroutines to improve performance that benefits everyone.

The Haxe coroutine implementation should not require any explicit yield or resume calls, instead leaving the compiler to mark when to yield and resume. Code quality will be better, and the behavior will mimic goroutines.

A default scheduler should be added into the Haxe stdlib, with the ability to shadow it for a custom implementation.

Given go2hx large test suite with a huge amount of benchmarks using coroutines, it will naturally put the project in a unique position in the future to improve the scheduler and coroutine implementations, there is no reason for this not to benefit all Haxe users as well.
 
The scheduler will also be written in Haxe in the stdlib and will be invoked at runtime where as the coroutine system will be written in OCAML in the compiler and will be invoked at compile time.

The coroutine implementation will add hooks to the scheduler to invoke it, which will be called at runtime.


## TLDR

- Elliott and I think that a coroutine and scheduler implementation in the Haxe compiler is the best outcome.
- I think that it's even better if it aligns as closely as possible with goroutines.
- Interoperability of go2hx and Haxe is highly valued and we should try to peruse it. 
- Is there any reason not to go down this route and see where it goes?

**Comment by 0b1kn00b** [comment]
The [Effect-TS](https://github.com/Effect-TS/effect) scheduler is interesting and worth looking at, imo.

**Comment by TheTechsTech** [comment]
Seems like https://github.com/profound7/moon-core?tab=readme-ov-file#generators is a good starting, many things that can be built upon or extracted.
